[SPARK-21333][Docs] Removed invalid joinTypes from javadoc of Dataset#joinWith

coreywoodfield opened the pull request:
## What changes were proposed in this pull request?

Two invalid join types were mistakenly listed in the javadoc for joinWith, in the Dataset class. I presume these were copied from the javadoc of join, but since joinWith returns a Dataset\<Tuple2\>, left_semi and left_anti are invalid, as they only return values from one of the datasets, instead of from both

## How was this patch tested?

I ran the following code : 
```
public static void main(String[] args) {
	SparkSession spark = new SparkSession(new SparkContext("local[*]", "Test"));
	Dataset<Row> one = spark.createDataFrame(Arrays.asList(new Bean(1), new Bean(2), new Bean(3), new Bean(4), new Bean(5)), Bean.class);
	Dataset<Row> two = spark.createDataFrame(Arrays.asList(new Bean(4), new Bean(5), new Bean(6), new Bean(7), new Bean(8), new Bean(9)), Bean.class);
		
	try {two.joinWith(one, one.col("x").equalTo(two.col("x")), "inner").show();} catch (Exception e) {e.printStackTrace();}
	try {two.joinWith(one, one.col("x").equalTo(two.col("x")), "cross").show();} catch (Exception e) {e.printStackTrace();}
	try {two.joinWith(one, one.col("x").equalTo(two.col("x")), "outer").show();} catch (Exception e) {e.printStackTrace();}
	try {two.joinWith(one, one.col("x").equalTo(two.col("x")), "full").show();} catch (Exception e) {e.printStackTrace();}
	try {two.joinWith(one, one.col("x").equalTo(two.col("x")), "full_outer").show();} catch (Exception e) {e.printStackTrace();}
	try {two.joinWith(one, one.col("x").equalTo(two.col("x")), "left").show();} catch (Exception e) {e.printStackTrace();}
	try {two.joinWith(one, one.col("x").equalTo(two.col("x")), "left_outer").show();} catch (Exception e) {e.printStackTrace();}
	try {two.joinWith(one, one.col("x").equalTo(two.col("x")), "right").show();} catch (Exception e) {e.printStackTrace();}
	try {two.joinWith(one, one.col("x").equalTo(two.col("x")), "right_outer").show();} catch (Exception e) {e.printStackTrace();}
	try {two.joinWith(one, one.col("x").equalTo(two.col("x")), "left_semi").show();} catch (Exception e) {e.printStackTrace();}
	try {two.joinWith(one, one.col("x").equalTo(two.col("x")), "left_anti").show();} catch (Exception e) {e.printStackTrace();}
}
```
which tests all the different join types, and the last two (left_semi and left_anti) threw exceptions. The same code using join instead of joinWith did fine. The Bean class was just a java bean with a single int field, x.

coreywoodfield committed:
Removed invalid joinTypes from javadoc of Dataset#joinWith

AmplabJenkins commented:
Can one of the admins verify this patch?

felixcheung commented:
how about checking if we have tests for these two types (as not supported)?

gatorsmile commented:
Yes. We should also capture it and throw an exception. For example, in line 1009, you can do something like

```Scala
    if (joined.joinType == LeftSemi || joined.joinType == LeftAnti) {
      throw new AnalysisException("XYZ")
    }
```

Then, you can add a test case in DatasetSuite.scala

coreywoodfield committed:
Added check for invalid join types in joinWith and tests

coreywoodfield commented:
I added the check as well as tests to make sure exceptions are thrown in the correct cases. If you want me to make the test more extensive (e.g. test that exceptions are not thrown when valid join types are passed in, etc.) I'd be happy to, just let me know

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:
```diff
@@ -321,6 +321,19 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
+    val ds1 = Seq(1, 2, 3).toDS().as("a")
+    val ds2 = Seq(1, 2).toDS().as("b")
+
+    intercept[AnalysisException] {
```
Yeah. Please check the output message. 
```
val e = intercept[AnalysisException] {
  ...
}.getMessages
assert(e.contains("xyz"))
```

gatorsmile commented:
Please create a JIRA in https://issues.apache.org/jira/projects/SPARK/summary and put the JIRA number in your PR title. You can follow http://spark.apache.org/contributing.html

coreywoodfield committed:
Added check to make sure correct exception was being caught in test

coreywoodfield commented:
I added a check on the output message and created a JIRA. I think that should cover everything. Thanks for all your guidance in this

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -1007,6 +1007,10 @@ class Dataset[T] private[sql](
         Some(condition.expr))).analyzed.asInstanceOf[Join]
 
+    if (joined.joinType == LeftSemi || joined.joinType == LeftAnti) {
+      throw new AnalysisException("Invalid join type in joinWith: " + joined.joinType)
```
Nit: `joined.joinType.sql`?

coreywoodfield committed:
Analysis exception now contains joinType.sql instead of joinType

Also made tests more robust and less likely to break if changes are made to joinTypes

gatorsmile commented:
retest this please

gatorsmile commented:
ok to test

SparkQA commented:
**[Test build #79354 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79354/testReport)** for PR 18462 at commit [`9aee54a`](https://github.com/apache/spark/commit/9aee54a0750bb3275aa2c48a50c12760cab15281).

SparkQA commented:
**[Test build #79354 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79354/testReport)** for PR 18462 at commit [`9aee54a`](https://github.com/apache/spark/commit/9aee54a0750bb3275aa2c48a50c12760cab15281).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79354/
Test PASSed.

coreywoodfield commented:
bump

gatorsmile commented:
sorry, I forgot it.

gatorsmile commented:
retest this please

gatorsmile commented:
LGTM

SparkQA commented:
**[Test build #79742 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79742/testReport)** for PR 18462 at commit [`9aee54a`](https://github.com/apache/spark/commit/9aee54a0750bb3275aa2c48a50c12760cab15281).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79742/
Test FAILed.

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #79767 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79767/testReport)** for PR 18462 at commit [`9aee54a`](https://github.com/apache/spark/commit/9aee54a0750bb3275aa2c48a50c12760cab15281).

SparkQA commented:
**[Test build #79767 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79767/testReport)** for PR 18462 at commit [`9aee54a`](https://github.com/apache/spark/commit/9aee54a0750bb3275aa2c48a50c12760cab15281).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79767/
Test PASSed.

gatorsmile commented:
Thanks! Merging to master/2.2
